web/course.go

```diff
@@ -94,15 +96,81 @@ func CoursePage(c *gin.Context) {
 		c.AbortWithStatus(http.StatusInternalServerError)
 		return
 	}
+
+	progs, err := dao.LoadProgressesForCourseAndUser((*tumLiveContext.Course).ID, (*tumLiveContext.User).ID)
+	if err != nil {
+		sentry.CaptureException(err)
+		log.Printf("%v", err)
+	}
+
+	progressResponses, progressStreams := prepareCourseProgressData(tumLiveContext, progs)
+
+	encoded, err := json.Marshal(progressResponses)
+	if err != nil {
+		sentry.CaptureException(err)
+		log.Printf("%v", err)
+	}
 	// in any other case assume either validated before or public course
-	err := templ.ExecuteTemplate(c.Writer, "course.gohtml", CoursePageData{IndexData: indexData, Course: *tumLiveContext.Course})
+	err = templ.ExecuteTemplate(c.Writer, "course-overview.gohtml",
+		CoursePageData{IndexData: indexData, Course: *tumLiveContext.Course, StreamsWithProgress: progressStreams, ProgressResponses: string(encoded)})
 	if err != nil {
 		sentrygin.GetHubFromContext(c).CaptureException(err)
 	}
 }
 
+func prepareCourseProgressData(tumLiveContext tools.TUMLiveContext, streamProgresses []model.StreamProgress) ([]ProgressInfo, []ProgressStream) {
+	var progressStreams []ProgressStream
+	var progressResponses []ProgressInfo
+
+	courseStreams := (*tumLiveContext.Course).Streams
+	// Combine streams with existing progresses.
+	for _, stream := range courseStreams {
```

Comment:
It seems a little like this function could be a (granted, rather large) sql statement, did you try that or do am I thinking in the wrong direction?

Reply:
No, this could be (probably) done in SQL. I just was not sure if this is worth the pain :joy: But I'll give it a try. :+1:

Reply:
Nah no worries, feel free to leave this as is 😀